Firmware/targets/arm/Src/encoder_hw.cpp

```diff
@@ -0,0 +1,122 @@
+#include "encoder_hw.h"
+
+#include "stm32f1xx_hal_gpio.h"
+#include "stm32f1xx_hal_tim.h"
+#include "tim.h"
+//#include "EncoderLogic.h"
+
+#define DEBOUNCE_STABLE 5
+#define ENCODER_PRESC 4
+
+void (*encoderCallback)() = NULL;
+
+volatile int encoderCount = 0;
+
+struct State {
+    bool p, n;
+};
+
+class LineDebounce {
+ public:
+    LineDebounce() {
+        state = UNSTABLE;
+        last_value = false;
+        times_stable = 0;
+    }
+    enum LineState { UNSTABLE, TRANSITION, STABLE };
+
+    void tick(bool now) {
+        switch (state) {
+            case UNSTABLE:
+                tickUnstable(now);
+                break;
+            case TRANSITION:
+                this->state = STABLE;
+                break;
+            case STABLE:
+                tickStable(now);
+                break;
+        }
+        last_value = now;
+    }
+    inline LineState getState() {
+        return state;
+    }
+    inline bool getValue() {
+        return last_value;
+    }
+
+ private:
+    LineState state;
+    bool last_value;
+    uint8_t times_stable;
+
+    void tickUnstable(bool now) {
+        if (now == last_value) {
+            ++times_stable;
+        } else {
+            times_stable = 0;
+        }
+        if (times_stable == DEBOUNCE_STABLE) {
+            this->state = TRANSITION;
+        }
+    }
+
+    void tickStable(bool now) {
+        if (now != last_value) {
+            times_stable = 0;
+            this->state = UNSTABLE;
+        }
+    }
+};
+
+void encoderInit(void) {
+    HAL_TIM_Base_Start_IT(&htim2);
+}
+
+int encoderGetAndReset(void) {
+    __disable_irq();
+    int toReturn = encoderCount / ENCODER_PRESC;
+    if (toReturn != 0)
+        encoderCount = 0;
```

Comment:
Nope, above line may cause double tick recognisation, ie:
1. after some ISR: encoderCount = 2  (ENCODER PRESC = 4)
2. (encoderCount %= ENCODER_PRESC) == 2 -> causes returning 2 and resetting
3. complete turn after some ISR: encoderCount = 2
4. Again returning 2...